agent/tool-scripts/postprocess/pidstat-convert

```diff
@@ -0,0 +1,114 @@
+        # 1547780606     0      2373    0.00    0.00    0.00    0.00     0      0.00      0.00  228772    7940   0.00      0.00      0.00      0.00       0      0.00      0.00  /usr/sbin/rsyslogd -n
+        # 1547780606     0      2380    0.00    0.00    0.00    0.00     7      0.00      0.00 1221772   15360   0.01      0.00      0.00      0.00       0      0.00      0.00  /usr/sbin/libvirtd
+        
+	###v11 with threads
+        # Time   UID      TGID       TID    %usr %system  %guest    %CPU   CPU  minflt/s  majflt/s     VSZ     RSS   %MEM   kB_rd/s   kB_wr/s kB_ccwr/s iodelay   cswch/s nvcswch/s  Command
+        # 1547852254     0         1         0    0.00    0.00    0.00    0.00     4      0.13      0.00  194520    7572   0.00      1.16     28.14      1.62      89      0.62      0.01  /usr/lib/systemd/systemd --switched-root --system --deserialize 22
+        # 1547852254     0         0         1    0.00    0.00    0.00    0.00     4      0.13      0.00  194520    7572   0.00      0.02      0.00      0.00      89      0.62      0.01  |__/usr/lib/systemd/systemd --switched-root --system --deserialize 22
+        #
+        if ( $line =~ /\s{0,1}(\d+)\s+\d+\s+(\S+|\S+\s+\S+)\s+(\d+\.\d+\s+\d+\.\d+\s+\d+\.\d+\s+\d+\.\d+\s+\d+\s+\d+\.\d+\s+\d+\.\d+\s+\d+\s+\d+\s+\d+\.\d+\s+\d+\.\d+\s+\d+\.\d+\s+\d+\.\d+\s+\d+\s+\d+\.\d+\s+\d+\.\d+\s+)(.*)/ ) {
+		my $time = $1;
+		my $pid = $2;
+		my $values_str = $3;
+		my $cmd = substr $4, 0, 40; # limit the length to 60 chars
+		my @values = split(/\s+/, $values_str);
+		if ($pid =~ /(^\S+)\s+(\S+)$/) {
+			my $tgid = $1;
+			my $tid = $2;
+			if ($tgid =~ /^0|\-$/) {
+				$pid = $prev_pid . "-" . $tid;
+				$cmd =~ s/\|__//;
+			} else {
+				$prev_pid = $tgid;
+				next; # Not going to save the per-process stats because we have the per-thread stats
+			}
+		}
+		# filenames have restrictions, and a process name with arguments don't always work as filename
+		
+		$cmd =~ s/%/%%/g; # avoid 'Invalid conversion in printf: "%=" ' errors if cmd contains a %.;
+		$cmd =~ s/\s$//g; # remove the last space pidstat always adds
+		$cmd =~ s/\s/_/g; # convert spaces to _
+		$cmd =~ s/\//_/g; # remove the first _
+		$cmd =~ s/\'//g;
+		$cmd =~ s/\"//g;
+		$cmd =~ s/^_//g;
+		$cmd =~ s/,//g; # commas cause problems in the CSV files.;
+		my $pid_cmd = $pid . ":" . $cmd;
+		my %pid_stats;
+		@pid_stats{@attributes} = @values;
+		$current_stats{$pid_cmd} = \%pid_stats;
+		# maintain a file for each of the PIDs instead of having all data in 1 file
+		my $file = $tool_output_dir . "/pids/" . $pid . "/" . $cmd;
+		if ( ! -e $file ) {
+			mkdir($tool_output_dir . "/pids/" . $pid);
+			open($file_handles{$pid}, ">", $file) || die "Could not open file $file";
+			select($file_handles{$pid});
+			$| = 1; # always flush
+			select(STDOUT);
+		}
+		if (exists $file_handles{$pid}) {
```

Comment:
What does it mean when this is `false`?

Doesn't line 89 mean that if it doesn't `die` there is a file handle in `file_handles{$pid}`?

And do we need to close the old file handle first?

Reply:
Originally I did not have this check, and for only a few pids/filehandles, I would get a "does not exist" for the filehandle, so I added this check.  It's unclear to me why the file-open would work, but I would not get a file handle.  I'll run a few more tests to see if we can find out why.

Reply:
I am not seeing this problem any more, and it may be for two reasons:
-I originally did not have ulimit -n included in the command to run pidstat-convert, and we may have hit the default limit of 1024 max open files
-When I was originally testing this, I cat the output of a previously run, pidstat stdout (3.2GB) to pidstat-convert, so it was not receiving new data every second, but as fast as it could process it.  Maybe there's an issue with opening files that quickly.